File: outreach/send.py
"""Approve + send (D2 AM): the human gate for Qualified -> Contacted.

do_not_send guard: unless OUTREACH_GO_LIVE=1, a recipient must be in DEMO_ALLOWED_RECIPIENTS
(comma-separated, .env) — this keeps demo sends inside inboxes the sourcer controls."""
from __future__ import annotations

import json
import os
import sqlite3
from typing import Any

from db.state import transition, utcnow


class SendBlocked(PermissionError):
    pass


def allowed_recipients() -> set[str]:
    raw = os.environ.get("DEMO_ALLOWED_RECIPIENTS") or os.environ.get("OUTREACH_ALLOWED_RECIPIENTS") or ""
    return {x.strip().lower() for x in raw.split(",") if x.strip()}


def go_live() -> bool:
    return os.environ.get("OUTREACH_GO_LIVE", "").strip().lower() in ("1", "true", "yes")


def is_allowed(to: str | None) -> bool:
    to = (to or "").strip().lower()
    return bool(to) and "@" in to and (go_live() or to in allowed_recipients())


def check_recipient(to: str) -> None:
    to = (to or "").strip().lower()
    if not to or "@" not in to:
        raise SendBlocked("recipient email is missing or invalid")
    if not is_allowed(to):
        raise SendBlocked(f"{to} is not in DEMO_ALLOWED_RECIPIENTS and OUTREACH_GO_LIVE is not set (do_not_send guard)")


def approve_and_send(conn: sqlite3.Connection, vendor_id: str, to: str, subject: str, body: str, gmail: Any,
                     draft_interaction_id: int | None = None, actor: str = "human") -> dict[str, Any]:
    """Send via the Gmail client, write interactions(direction=outbound, gmail_thread_id), transition to Contacted."""
    check_recipient(to)
    v = conn.execute("SELECT status FROM vendors WHERE vendor_id = ?", (vendor_id,)).fetchone()
    if v is None:
        raise KeyError(vendor_id)
    if v["status"] != "Qualified":
        raise SendBlocked(f"vendor is {v['status']}, only Qualified vendors can be contacted")
    if not subject.strip() or not body.strip():
        raise SendBlocked("subject and body are required")
    sent = gmail.send(to=to, subject=subject, body=body)
    meta = {"gmail_message_id": sent["message_id"], "draft_interaction_id": draft_interaction_id, "to": to}
    with conn:
        cur = conn.execute(
            "INSERT INTO interactions (vendor_id, gmail_thread_id, direction, sent_at, subject, body_text, llm_summary) "
            "VALUES (?, ?, 'outbound', ?, ?, ?, ?)",
            (vendor_id, sent["thread_id"], utcnow(), subject, body, json.dumps(meta)))
        conn.execute("UPDATE vendors SET contact_email = ?, next_action = 'await_reply', updated_at = ? WHERE vendor_id = ?",
                     (to, utcnow(), vendor_id))
    interaction_id = int(cur.lastrowid)
    event_id = transition(vendor_id, "Contacted", actor, reason=f"outreach email sent to {to}",
                          evidence_ref=interaction_id, payload={"gmail_thread_id": sent["thread_id"]}, conn=conn)
    return {"interaction_id": interaction_id, "event_id": event_id, **sent}

File: src/lib/outreach/send.ts
/**
 * The human send gate (PRD §6 Qualified -> Contacted), as a domain operation:
 * allowlist, status rules, thread reuse for follow-ups, the Gmail call, then
 * the interaction row, the transition and the vendor's next action. The route
 * only validates input and maps SendError codes to HTTP statuses.
 */
import { and, desc, eq } from "drizzle-orm";

import { getDb, interactions, nowIso, vendors, type Db } from "@/lib/db";
import { getVendorDetail, insertInteraction } from "@/lib/db/queries";
import type { InteractionRow } from "@/lib/db/schema";
import { TransitionError, transition, type TransitionResult } from "@/lib/db/state";

import { isAllowedRecipient, normalizeEmail, parseAllowlist } from "./allowlist";
import { GmailNotConfiguredError, GmailNotConnectedError, profile, sendMessage } from "./gmail";

/** A follow-up is expected within a week; the Board shows the date and flags it when overdue. */
export const FOLLOW_UP_DAYS = 7;

export type SendInput = { to: string; subject: string; body: string; draft_interaction_id?: number };

export type SendErrorCode =
  | "recipient_not_allowed"
  | "vendor_not_found"
  | "wrong_status"
  | "no_thread"
  | "gmail_not_connected"
  | "gmail_failed"
  | "recording_failed";

export class SendError extends Error {
  constructor(
    public readonly code: SendErrorCode,
    message: string,
    public readonly details: Record<string, unknown> = {},
  ) {
    super(message);
  }
}

export function sendErrorStatus(code: SendErrorCode): number {
  switch (code) {
    case "recipient_not_allowed":
      return 403;
    case "vendor_not_found":
      return 404;
    case "gmail_failed":
      return 502;
    case "recording_failed":
      return 500;
    default:
      return 409;
  }
}

/** Seam for tests and for a different mail provider later. */
export type Mailer = {
  profile: () => Promise<{ email: string }>;
  sendMessage: (to: string, subject: string, body: string, opts?: { threadId?: string }) => Promise<{ threadId: string; messageId: string }>;
};

const gmailMailer: Mailer = { profile, sendMessage };

export type SendResult = {
  thread_id: string;
  message_id: string;
  sender: string;
  follow_up: boolean;
  interaction: InteractionRow;
  transition: TransitionResult | null;
};

export async function sendOutreach(
  vendorId: string,
  input: SendInput,
  deps: { db?: Db; mailer?: Mailer; allowlist?: string[] } = {},
): Promise<SendResult> {
  const db = deps.db ?? getDb();
  const mailer = deps.mailer ?? gmailMailer;
  const allowlist = deps.allowlist ?? parseAllowlist();
  const to = normalizeEmail(input.to);

  if (!isAllowedRecipient(to, allowlist)) {
    throw new SendError(
      "recipient_not_allowed",
      allowlist.length ? `${to} is not in DEMO_ALLOWED_RECIPIENTS` : "DEMO_ALLOWED_RECIPIENTS is empty; add allowed addresses to .env.local before sending",
      { allowed_recipients: allowlist },
    );
  }
  const detail = getVendorDetail(vendorId, db);
  if (!detail) throw new SendError("vendor_not_found", `vendor ${vendorId} not found`);

  // First contact leaves Qualified; a Contacted or Dormant vendor gets an in-thread follow-up with no status change.
  const followUp = detail.vendor.status === "Contacted" || detail.vendor.status === "Dormant";
  if (detail.vendor.status !== "Qualified" && !followUp) {
    throw new SendError("wrong_status", `vendor is ${detail.vendor.status}; only Qualified vendors can be contacted (Contacted/Dormant get in-thread follow-ups)`);
  }
  const priorThread = followUp
    ? db.select().from(interactions).where(and(eq(interactions.vendor_id, vendorId), eq(interactions.direction, "outbound"))).orderBy(desc(interactions.interaction_id)).get()
    : undefined;
  if (followUp && !priorThread?.gmail_thread_id) throw new SendError("no_thread", "no earlier outbound thread to follow up on");

  let sender: string;
  let sent: { threadId: string; messageId: string };
  try {
    sender = (await mailer.profile()).email;
    sent = await mailer.sendMessage(to, input.subject, input.body, followUp ? { threadId: priorThread?.gmail_thread_id ?? undefined } : {});
  } catch (err) {
    if (err instanceof GmailNotConnectedError || err instanceof GmailNotConfiguredError) {
      throw new SendError("gmail_not_connected", err.message, { auth_url: "/api/gmail/auth" });
    }
    throw new SendError("gmail_failed", `Gmail send failed: ${err instanceof Error ? err.message : String(err)}`);
  }

  try {
    const now = nowIso();
    return db.transaction((tx) => {
      const interaction = insertInteraction(
        {
          vendor_id: vendorId,
          direction: "outbound",
          gmail_thread_id: sent.threadId,
          sent_at: now,
          subject: input.subject,
          body_text: input.body,
          llm_summary: `${followUp ? "follow-up" : "first contact"}${input.draft_interaction_id ? ` from draft #${input.draft_interaction_id}` : ""}`,
        },
        tx,
      );
      const t = followUp
        ? null
        : transition(
            {
              vendorId,
              toStatus: "Contacted",
              actor: "human",
              reason: "sent outreach email",
              evidenceRef: `interaction:${interaction.interaction_id}`,
              payload: { gmail_thread_id: sent.threadId, gmail_message_id: sent.messageId, to, subject: input.subject },
            },
            tx,
          );
      tx.update(vendors)
        .set({
          owner: sender || detail.vendor.owner,
          contact_email: detail.vendor.contact_email ?? to,
          next_action: "await_reply",
          due_at: new Date(Date.now() + FOLLOW_UP_DAYS * 86_400_000).toISOString(),
          updated_at: now,
        })
        .where(eq(vendors.vendor_id, vendorId))
        .run();
      return { thread_id: sent.threadId, message_id: sent.messageId, sender, follow_up: followUp, interaction, transition: t };
    });
  } catch (err) {
    // The email is out; make the inconsistency loud rather than silent.
    console.error(`[send ${vendorId}] email sent (thread ${sent.threadId}) but recording failed`, err);
    throw new SendError(
      "recording_failed",
      `email sent (thread ${sent.threadId}) but ${err instanceof TransitionError ? "the status change" : "recording"} failed: ${err instanceof Error ? err.message : String(err)}`,
      { thread_id: sent.threadId },
    );
  }
}
